### https://blog.ippon.tech/mongodb-and-apache-spark-getting-started-tutorial/

In [1]:
import sys, os
import pprint
#from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql.window import Window
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql import dataframe
from pyspark.sql.dataframe import DataFrame
#--------------------------------------------------------
#os.environ['Java_HOME']='C:\xxxx'
'''
try:
    sc = SparkContext()
except:
    print("Spark Context exists")
sc.version
'''

'\ntry:\n    sc = SparkContext()\nexcept:\n    print("Spark Context exists")\nsc.version\n'

In [2]:
def transform(self, f):
    return f(self)

DataFrame.transform = transform
#dir(DataFrame)

In [ ]:
'''
working_directory = './mongo-jars/*'
spark = SparkSession.builder.appName("mongoApp") \
    .config("spark.mongodb.input.uri", "mongodb://ds147480.mlab.com:47480/deliverdb.hello") \
    .config("spark.mongodb.output.uri", "mongodb://ds147480.mlab.com:47480/deliverdb.hello") \
    .config('spark.jars.packages', 'org.mongodb.spark:mongo-spark-connector_2.11:2.4.1') \
    .config('spark.driver.extraClassPath', working_directory) \
    .getOrCreate()
'''

## If using SparkContext, then below:

In [ ]:
#If using SparkContext, then below:
from pyspark import SparkContext,SparkConf
conf = SparkConf().set("spark.jars.packages", "org.mongodb.spark:mongo-spark-connector_2.11:2.4.1")
sc = SparkContext(conf=conf)

from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("myApp") \
.config("spark.mongodb.input.uri", "mongodb://suman1:suman1@ds147480.mlab.com:47480/deliverdb.members") \
.config("spark.mongodb.output.uri", "mongodb://suman1:suman1@ds147480.mlab.com:47480/deliverdb.members") \
.getOrCreate()

df = spark.read.format("com.mongodb.spark.sql.DefaultSource").load()
df.printSchema()

## If using SparkSession, then below:

In [3]:
#If using SparkSession, then below:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("myApp") \
    .config("spark.mongodb.input.uri", "mongodb://suman1:suman1@ds147480.mlab.com:47480/deliverdb.members") \
    .config("spark.mongodb.output.uri", "mongodb://suman1:suman1@ds147480.mlab.com:47480/deliverdb.members") \
    .config('spark.jars.packages', 'org.mongodb.spark:mongo-spark-connector_2.11:2.4.1') \
    .getOrCreate()

In [ ]:
df = spark.read.format("com.mongodb.spark.sql.DefaultSource").load()
df2 = spark.read.format("com.mongodb.spark.sql.DefaultSource").option("uri","mongodb://suman1:suman1@ds147480.mlab.com:47480/deliverdb.members").load()
df.printSchema()
df2.toPandas()

In [ ]:
#df2.select("coverages"(0)("type")).toPandas()
df2.select("coverages.type", "coverages.claim_amt").toPandas()
#df2.select("coverages.type.0").toPandas()

In [ ]:
df2.select("coverages").toPandas()

### Start Hive/CSV/RDBMS to MongoB Ingestion POC

In [4]:
emp_df = spark.read.csv('data/emp.txt', header=True)
emp_df.toPandas()

emp_id  emp_name
0    111     Suman
1    222  Souranya
2    333      Saby
3    444    Vinesh

In [5]:
adrs_df = spark.read.csv('data/adrs.txt', header=True)
adrs_df.toPandas()

emp_id adrs_type            adrs_line_1
0    111      Home         5250 Town Blvd
1    111      Work          8283 Plano Rd
2    222      Home         5250 Town Blvd
3    333      Home  Somewhere in Tenessey
4    333      Work   Some ofc in Tenessey

In [6]:
phone_df = spark.read.csv('data/phone.txt', header=True)
phone_df.toPandas()

emp_id ph_type      ph_num
0    111    Home  5133028287
1    111    Work  1113334444
2    222    Home  2224853691
3    333    Home  3334859127
4    333    Work  3326541298

### Join Member & Address

In [7]:
e = emp_df.alias('e')
a = adrs_df.alias('a')
mbr_adrs_df = e.join(a, (e.emp_id == a.emp_id), how='inner')
mbr_adrs_df = mbr_adrs_df.drop(a.emp_id)
mbr_adrs_df.toPandas()

emp_id  emp_name adrs_type            adrs_line_1
0    111     Suman      Home         5250 Town Blvd
1    111     Suman      Work          8283 Plano Rd
2    222  Souranya      Home         5250 Town Blvd
3    333      Saby      Home  Somewhere in Tenessey
4    333      Saby      Work   Some ofc in Tenessey

### Join Member & Phone

In [8]:
p = phone_df.alias('p')
mbr_ph_df = e.join(p, (e.emp_id == p.emp_id), how='inner')
mbr_ph_df = mbr_ph_df.drop(p.emp_id)
mbr_ph_df.toPandas()

emp_id  emp_name ph_type      ph_num
0    111     Suman    Home  5133028287
1    111     Suman    Work  1113334444
2    222  Souranya    Home  2224853691
3    333      Saby    Home  3334859127
4    333      Saby    Work  3326541298

## https://stackoverflow.com/questions/52142272/pyspark-dataframe-creating-a-nested-structure

In [9]:
#Build Member & Address Struct
#joined_df = joined_df.groupBy('emp_id','emp_name').agg(collect_list('adrs_line_1').alias('address'))
mbr_adrs_df = mbr_adrs_df.groupBy('emp_id','emp_name').agg(collect_list(struct(['adrs_type','adrs_line_1'])).alias('address'))
mbr_adrs_df.toPandas()

emp_id  emp_name                                            address
0    222  Souranya                           [(Home, 5250 Town Blvd)]
1    333      Saby  [(Home, Somewhere in Tenessey), (Work, Some of...
2    111     Suman    [(Home, 5250 Town Blvd), (Work, 8283 Plano Rd)]

In [10]:
#Build Member & Phone Struct
mbr_ph_df2 = mbr_ph_df.groupBy('emp_id','emp_name').agg(collect_list(struct(['ph_type','ph_num'])).alias('phone'))
mbr_ph_df2.toPandas()

emp_id  emp_name                                     phone
0    222  Souranya                      [(Home, 2224853691)]
1    333      Saby  [(Home, 3334859127), (Work, 3326541298)]
2    111     Suman  [(Home, 5133028287), (Work, 1113334444)]

### Desired JSON layout:

> {  
   "emp_id":111,  
   "emp_name":"Suman",  
   "address":[  
	{  
      "adrs_type":"Home",  
      "adrs_line_1":"5250 Town Blvd"  
	}  
   ]  
}

In [11]:
pp = pprint.PrettyPrinter(indent=4)
for j in mbr_adrs_df.toJSON().collect():
    print(j)
    print('\n')

{"emp_id":"222","emp_name":"Souranya","address":[{"adrs_type":"Home","adrs_line_1":"5250 Town Blvd"}]}


{"emp_id":"333","emp_name":"Saby","address":[{"adrs_type":"Home","adrs_line_1":"Somewhere in Tenessey"},{"adrs_type":"Work","adrs_line_1":"Some ofc in Tenessey"}]}


{"emp_id":"111","emp_name":"Suman","address":[{"adrs_type":"Home","adrs_line_1":"5250 Town Blvd"},{"adrs_type":"Work","adrs_line_1":"8283 Plano Rd"}]}




### Convert Spark DF with 2D struct into nested JSON

In [12]:
for j in mbr_ph_df2.toJSON().collect():
    print(j)
    print('\n')

{"emp_id":"222","emp_name":"Souranya","phone":[{"ph_type":"Home","ph_num":"2224853691"}]}


{"emp_id":"333","emp_name":"Saby","phone":[{"ph_type":"Home","ph_num":"3334859127"},{"ph_type":"Work","ph_num":"3326541298"}]}


{"emp_id":"111","emp_name":"Suman","phone":[{"ph_type":"Home","ph_num":"5133028287"},{"ph_type":"Work","ph_num":"1113334444"}]}




### Build the final joined dataframe with Member, Addresses & Phones

In [33]:
ma = mbr_adrs_df.alias('ma')
mp = mbr_ph_df2.alias('mp')
final_df = ma.join(mp, (ma.emp_id == mp.emp_id), how='inner')
final_df=final_df.drop(ma.emp_id).drop(ma.emp_name)
final_df.printSchema()
final_df2 = final_df.select('emp_id','emp_name','address','phone')
final_df2.toPandas()

root
 |-- address: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- adrs_type: string (nullable = true)
 |    |    |-- adrs_line_1: string (nullable = true)
 |-- emp_id: string (nullable = true)
 |-- emp_name: string (nullable = true)
 |-- phone: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- ph_type: string (nullable = true)
 |    |    |-- ph_num: string (nullable = true)



emp_id  emp_name                                            address  \
0    222  Souranya                           [(Home, 5250 Town Blvd)]   
1    333      Saby  [(Home, Somewhere in Tenessey), (Work, Some of...   
2    111     Suman    [(Home, 5250 Town Blvd), (Work, 8283 Plano Rd)]   

                                      phone  
0                      [(Home, 2224853691)]  
1  [(Home, 3334859127), (Work, 3326541298)]  
2  [(Home, 5133028287), (Work, 1113334444)]

### Convert Spark DF with multiple 2D structs to JSON

In [16]:
for j in final_df2.toJSON().collect():
    print(j)
    print('\n')

{"emp_id":"222","emp_name":"Souranya","address":[{"adrs_type":"Home","adrs_line_1":"5250 Town Blvd"}],"phone":[{"ph_type":"Home","ph_num":"2224853691"}]}


{"emp_id":"333","emp_name":"Saby","address":[{"adrs_type":"Home","adrs_line_1":"Somewhere in Tenessey"},{"adrs_type":"Work","adrs_line_1":"Some ofc in Tenessey"}],"phone":[{"ph_type":"Home","ph_num":"3334859127"},{"ph_type":"Work","ph_num":"3326541298"}]}


{"emp_id":"111","emp_name":"Suman","address":[{"adrs_type":"Home","adrs_line_1":"5250 Town Blvd"},{"adrs_type":"Work","adrs_line_1":"8283 Plano Rd"}],"phone":[{"ph_type":"Home","ph_num":"5133028287"},{"ph_type":"Work","ph_num":"1113334444"}]}




In [21]:
df = final_df
df = df.drop(ma.emp_id)
df.printSchema()
df.select('emp_id').toPandas()

root
 |-- emp_name: string (nullable = true)
 |-- address: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- adrs_type: string (nullable = true)
 |    |    |-- adrs_line_1: string (nullable = true)
 |-- emp_id: string (nullable = true)
 |-- emp_name: string (nullable = true)
 |-- phone: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- ph_type: string (nullable = true)
 |    |    |-- ph_num: string (nullable = true)



emp_id
0    222
1    333
2    111